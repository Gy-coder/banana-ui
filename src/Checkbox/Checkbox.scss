.g-checkbox {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  color: #000000d9;
  font-size: 14px;
  list-style: none;
  position: relative;
  top: 0.2em;
  display: inline-block;
  outline: none;
  cursor: pointer;
  &-wrapper {
    box-sizing: border-box;
    margin: 0 8px 0 0;
    padding: 0;
    font-size: 14px;
    line-height: 1.5715;
    list-style: none;
    position: relative;
    display: inline-flex;
    align-items: baseline;
    cursor: pointer;
    &.checked {
      > .g-checkbox {
        > .g-checkbox-inner {
          background-color: #049aff;
          border-color: #0190fe;
        }
        > .g-checkbox-inner::after {
          display: block;
          box-sizing: content-box;
          content: '';
          border: 2px solid #fff;
          border-left: 0;
          border-top: 0;
          height: 8px;
          left: 5px;
          position: absolute;
          top: 1px;
          transform: rotate(45deg);
          width: 4px;
          transition: transform 0.15s cubic-bezier(0.71, -0.46, 0.88, 0.6) 0.05s;
          transform-origin: center;
        }
      }
    }
    &.disabled {
      cursor: not-allowed;
      > .g-checkbox {
        cursor: not-allowed;
        > .g-checkbox-inner {
          cursor: not-allowed;
          background-color: #eef1f6;
          border-color: #d1dbe5;
        }
        > input {
          cursor: not-allowed;
        }
      }
      &.checked {
        &.disabled {
          > .g-checkbox {
            > .g-checkbox-inner::after {
              display: block;
              box-sizing: content-box;
              content: '';
              border: 2px solid #b8b8b8;
              border-left: 0;
              border-top: 0;
              height: 8px;
              left: 5px;
              position: absolute;
              top: 1px;
              transform: rotate(45deg);
              width: 4px;
              transition: transform 0.15s cubic-bezier(0.71, -0.46, 0.88, 0.6)
                0.05s;
              transform-origin: center;
            }
          }
        }
      }
    }
  }

  &-wrapper + &-wrapper {
    margin-left: 8px;
  }
  > input {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    z-index: 1;
    cursor: pointer;
    opacity: 0;
  }
  &-inner {
    display: inline-block;
    position: relative;
    border: 1px solid #bfcbd9;
    border-radius: 4px;
    box-sizing: border-box;
    width: 18px;
    height: 18px;
    background-color: #fff;
    z-index: 1;
    transition: border-color 0.25s cubic-bezier(0.71, -0.46, 0.29, 1.46),
      background-color 0.25s cubic-bezier(0.71, -0.46, 0.29, 1.46);
  }
  &-tips {
    margin-left: 8px;
  }
}

.g-checkbox-group.vertical {
  display: flex;
  flex-direction: column;
  .g-checkbox-wrapper + .g-checkbox-wrapper {
    margin-left: 0;
  }
}
