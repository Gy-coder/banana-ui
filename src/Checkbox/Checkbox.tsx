import React from 'react';
import classnames from 'classnames';
import './Checkbox.scss';

export interface Props {
  checked: boolean;
  value: string;
  onChange: (value: string, e: React.ChangeEvent<HTMLInputElement>) => void;
  disabled?: boolean;
}

const Checkbox: React.FC<Props> = (props) => {
  const { children, value, checked, onChange, disabled = false } = props;
  const handleChange = (
    value: string,
    e: React.ChangeEvent<HTMLInputElement>,
  ) => {
    onChange(value, e);
  };
  const classes = classnames('g-checkbox-wrapper', {
    disabled,
    checked,
  });
  return (
    <label className={classes}>
      <span className="g-checkbox">
        <span className="g-checkbox-inner"></span>
        <input
          type="checkbox"
          value={value}
          onChange={(e) => handleChange(value, e)}
          checked={checked}
          disabled={disabled}
        />
      </span>
      <span className="g-checkbox-tips">{children}</span>
    </label>
  );
};

export default Checkbox;
