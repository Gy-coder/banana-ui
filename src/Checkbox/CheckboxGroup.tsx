import React from 'react';
import classnames from 'classnames';
//@ts-ignore
import Checkbox, { Props as CheckboxProps } from './Checkbox';

interface Props {
  values: string[];
  onChange: (values: string[]) => void;
  disabled: boolean;
}

const CheckboxGroup: React.FC<Props> = (props) => {
  const { disabled, onChange, values, children } = props;
  const classes = classnames('g-checkbox-group', {
    disabled,
  });
  const handleChange = (
    value: string,
    e: React.ChangeEvent<HTMLInputElement>,
  ) => {
    let valuesCopy = JSON.parse(JSON.stringify(values));
    if (!e.target.checked) {
      let index = values.indexOf(value);
      valuesCopy.splice(index, 1);
    } else {
      valuesCopy.push(value);
    }
    onChange(valuesCopy);
  };
  const render = () => {
    return React.Children.map(children, (child) => {
      const childElement =
        child as React.FunctionComponentElement<CheckboxProps>;
      if (childElement.type !== Checkbox) {
        throw new Error('CheckboxGroup的子元素必须是Checkbox');
      }
      return React.cloneElement(childElement, {
        checked: values.indexOf(childElement.props.value) >= 0,
        onChange: handleChange,
        disabled: childElement.props.disabled || disabled,
      });
    });
  };
  return <div className={classes}>{render()}</div>;
};

export default CheckboxGroup;
