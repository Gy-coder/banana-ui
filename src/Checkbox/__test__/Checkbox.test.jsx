import {
  act,
  render,
  cleanup,
  screen,
  fireEvent,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import React, { useState } from 'react';
import Checkbox from '../Checkbox';

afterEach(cleanup);

describe('测试CheckBox', () => {
  it('CheckBox可以正常渲染', () => {
    const { asFragment } = render(<Checkbox value="123">My Checkbox</Checkbox>);
    expect(screen.getByText(/My Checkbox/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
  it('可以disabled', () => {
    render(
      <Checkbox value="123" disabled={true}>
        My Checkbox
      </Checkbox>,
    );
    const el = screen.getByDisplayValue('123');
    expect(el).toBeDisabled();
  });
  it('可以改变值', () => {
    const Comp = () => {
      const [checked, setChecked] = useState(false);

      return (
        <div data-testid="checkbox">
          <Checkbox
            value="one"
            checked={checked}
            onChange={() => {
              setChecked(!checked);
            }}
          >
            Option One
          </Checkbox>
        </div>
      );
    };
    render(<Comp />);

    const el = screen.getByTestId('checkbox');
    const label = el.childNodes[0];
    expect(label.className).not.toContain('checked');
    fireEvent.click(label);
    expect(label.className).toContain('checked');
  });
});
