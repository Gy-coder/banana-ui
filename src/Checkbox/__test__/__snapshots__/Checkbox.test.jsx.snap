// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试CheckBox CheckBox可以正常渲染 1`] = `
<DocumentFragment>
  <label
    class="g-checkbox-wrapper"
  >
    <span
      class="g-checkbox"
    >
      <span
        class="g-checkbox-inner"
      />
      <input
        type="checkbox"
        value="123"
      />
    </span>
    <span
      class="g-checkbox-tips"
    >
      My Checkbox
    </span>
  </label>
</DocumentFragment>
`;
